data/blog/designing-sentrys-cross-region-replication.md

```diff
@@ -0,0 +1,160 @@
+
+# Evaluating existing solutions
+
+With a better understanding of our requirements, and workflows, we began evaluating existing solutions for how much complexity and infrastructure they would require in addition to meeting our functional requirements.
+
+## Postgres Replication
+
+Postgres comes with built-in capabilities to stream data from one database server into a replica that supports read operations. Postgres replication would have met our consistency requirements and would have required the smallest number of application changes. Postgres replication was not selected for a few reasons:
+
+1. Postgres can only replicate at the table level. This would result in data being ‘over-replicated’ to regions. If a US organization creates an Authentication Token, there is no reason for that token to be replicated to other regions.
+2. Postgres replication could not be used for the Region → Control path as a replica cannot have multiple primaries/leaders, and with each region replicating data back to control, we would have that scenario.
+
+### Change Data Capture (CDC) and Kafka
+
+Because naive Postgres replication would have resulted in over-replication, we could build an application that consumed Postgres’ streaming replication data, and converted changes into Kafka messages. A Postgres replication consumer would allow us to selectively replicate changes to only the regions that were relevant, and transform queries into relevant domain actions when creating Kafka messages.
+
+![flowchart of the cdc design](/images/designing-sentrys-cross-region-replication/cdc-kafka.png)
+
+CDC would provide atomic operations and we could easily backlog operations in Kafka for as long as required. Regions would use [mirror maker](https://developers.redhat.com/articles/2023/11/13/demystifying-kafka-mirrormaker-2-use-cases-and-architecture#use_cases) to replicate topics, or use a [multi-region Kafka](https://docs.confluent.io/platform/current/multi-dc-deployments/multiregion.html)
+
+We decided against this approach because of the complexity. We would need to operate additional Kafka clusters, maintain and operate a WAL consumer, Kafka producers, consumers, and translation code from WAL operations into replication changes.
+
+# Outboxes
+
+[Transactional Outboxes](https://microservices.io/patterns/data/transactional-outbox.html) are a distributed systems pattern that fit our use case perfectly. As the application makes changes that need to be replicated, it can save an ‘outbox’ in the same postgres transaction as the change - providing use the atomicity we wanted. In the background, a worker pulls tasks from the outbox table and runs handlers to apply the necessary replication action via RPC. With this design we would be able to provide eventual consistency with at-least-once delivery semantics.
```

Comment:
is _providing use the_ intentional here?

```suggestion
[Transactional Outboxes](https://microservices.io/patterns/data/transactional-outbox.html) are a distributed systems pattern that fit our use case perfectly. As the application makes changes that need to be replicated, it can save an ‘outbox’ in the same postgres transaction as the change - providing the atomicity we wanted. In the background, a worker pulls tasks from the outbox table and runs handlers to apply the necessary replication action via RPC. With this design we would be able to provide eventual consistency with at-least-once delivery semantics.
```

Reply:
nope, that is a mistake.